### PR TITLE
changed the linting for NextJS project using husky

### LINKED ISSUE #51: Better pre-commit config for nextjs project
following  https://github.com/memfreeme/memfree/issues/48

### PR DESCRIPTION
**Pull request by sank8-2**
Added husky, lint for nextjs and prettier for code formatting

### How to setup
Run ```npm i``` from the root directory, all packages and linters gets installed and setup in the local

Then it runs automatically at every commit without extra configuration.

Closes #51

### CODE REVIEW

**Comment by sank8-2** [comment]
@ahaapple Now you can check it, this one is better and suits a NextJS repo.
Although I haven't checked all the files for linting and formatting

File: .prettierrc [hunk 1]
```diff
@@ -0,0 +1,10 @@
+{
+    "bracketSpacing": true,
+    "endOfLine": "lf",
+    "printWidth": 80,
+    "semi": true,
+    "singleQuote": true,
```
Comment (ahaapple) [inline .prettierrc:6]: Hello. please this config content consistent with the  frontend/.prettierrc

File: package.json [hunk 2]
```diff
@@ -1,9 +1,26 @@
 {
   "dependencies": {
-    "husky": "^9.1.6"
+    "@typescript-eslint/eslint-plugin": "^8.10.0",
+    "@typescript-eslint/parser": "^8.10.0",
+    "eslint": "^8.57.1",
```
Comment (ahaapple) [inline package.json:5]: I think we could move all these files to the frontent directory.  This way we don't have to repeat a lot of things

File: .prettierignore [hunk 3]
```diff
@@ -0,0 +1,8 @@
+.next
+.cache
+package-lock.json
+public
+node_modules
```
Comment (ahaapple) [inline .prettierignore:5]: we could use frontend/.prettierignore

File: .eslintrc.js [hunk 4]
```diff
@@ -0,0 +1,67 @@
+module.exports = {
+    parser: '@typescript-eslint/parser',
+    parserOptions: {
+      ecmaVersion: 2021,
+      sourceType: 'module',
+      ecmaFeatures: {
+        jsx: true,
+      },
+    },
+    env: {
```
Comment (ahaapple) [inline .eslintrc.js:10]: we could use  frontend/.eslintrc.json

**Comment by sank8-2** [comment]
On it

File: package.json (hunk 2 above)
Comment (sank8-2) [reply package.json:5]: We can do that, but then only the frontend files will be checked.
And the way husky works is it checks for the staged files during pre-commit.
As git is initialized at the root directory, we have to run linters from there which will require these packages to be present at the root.

The one thing that can be done is move the repeated files from the frontend to the base, as they get checked anyhow from the pre-commit

Comment (ahaapple) [reply package.json:5]: 1. Only check the frontend project is enough.
2. We should keep the root directory content to be clean
3. We can first let the pre commit check work on the frontend and actually run it

Comment (sank8-2) [reply package.json:5]: Got your point, but the way husky works is from a git initialized location.
1. If aiming for frontend, then git has to be initialized there, then we can run husky from there

**Comment by sank8-2** [comment]
@ahaapple I have done the necessary changes, please review now

**Comment by ahaapple** [comment]
1. There are a lot of duplicate files, which is unreasonable
2. The contents of the root directory are too messy

**Comment by ahaapple** [comment]
I think a more elegant solution may be to use GitHub Actions to check each PR to be merged. Developers can use VScode to solve TS checks, code quality checks, and code formatting during the development process. We only need to check once before merging the PR. The benefits of this are:

1 Developers do not need to perform additional operations locally, which is more developer-friendly
2 The code repository will be neater

**Comment by sank8-2** [comment]
Ok, I'll discard the husky and add it to GitHub Actions

**Review by ahaapple — APPROVED** [review]

**Comment by ahaapple** [comment]
```
Run npm run lint
  npm run lint
  shell: /usr/bin/bash -e {0}

> @ lint /home/runner/work/memfree/memfree/frontend
> next lint

You are using Node.js 14.[2](https://github.com/memfreeme/memfree/actions/runs/11513014014/job/32048965546?pr=54#step:5:2)1.3. For Next.js, Node.js version >= v18.17.0 is required.
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! @ lint: `next lint`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the @ lint script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/runner/.npm/_logs/202[4](https://github.com/memfreeme/memfree/actions/runs/11513014014/job/32048965546?pr=54#step:5:5)-10-2[5](https://github.com/memfreeme/memfree/actions/runs/11513014014/job/32048965546?pr=54#step:5:6)T06_24_49_556Z-debug.log
Error: Process completed with exit code 1.
```
seems  we need to update the node version:
```
jobs:
  build:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v3
      - name: Use Node.js 18
        uses: actions/setup-node@v3
        with:
          node-version: '18.x'

```

**Comment by sank8-2** [comment]
@ahaapple Basically its performing linting and prettier check for the frontend directory
Are any other actions required?

**Comment by ahaapple** [comment]
@sank8-2  Thanks for your contribution. I think the current actions are enough. When we have enough test cases, we can add the check action of `npm run test` later.

**Comment by sank8-2** [comment]
@ahaapple Ok
Thank you

**Pull request merged**